File: packages/template-php/src/Value/Number.php
<?php

declare(strict_types=1);

namespace Polyspec\Template\Value;

/**
 * Number conversion and formatting (VAL-9, VAL-10, FUN-20 to FUN-25).
 */
final class Number
{
    public const MAX_SAFE = 9007199254740991;

    /**
     * VAL-9: ECMAScript Number::toString.
     */
    public static function toText(float $value): string
    {
        if ($value == 0.0) {
            return '0';
        }
        [$negative, $digits, $exponent] = self::shortestDigits($value);
        $k = strlen($digits);
        $n = $exponent;
        if ($k <= $n && $n <= 21) {
            $text = $digits . str_repeat('0', $n - $k);
        } elseif (0 < $n && $n <= 21) {
            $text = substr($digits, 0, $n) . '.' . substr($digits, $n);
        } elseif (-6 < $n && $n <= 0) {
            $text = '0.' . str_repeat('0', -$n) . $digits;
        } else {
            $e = $n - 1;
            $sign = $e < 0 ? '-' : '+';
            $mantissa = $k === 1 ? $digits : $digits[0] . '.' . substr($digits, 1);
            $text = $mantissa . 'e' . $sign . abs($e);
        }

        return $negative ? '-' . $text : $text;
    }

    /**
     * Shortest round-trip digits and the exponent n such that value = 0.digits × 10^n.
     *
     * @return array{0: bool, 1: string, 2: int}
     */
    public static function shortestDigits(float $value): array
    {
        $negative = $value < 0;
        $abs = abs($value);
        // serialize_precision -1 gives the shortest round-trip representation.
        $text = (string) json_encode($abs, JSON_PRESERVE_ZERO_FRACTION);
        if (!preg_match('/^(\d+)(?:\.(\d+))?(?:[eE]([+-]?\d+))?$/', $text, $m)) {
            throw new \RuntimeException("cannot decompose {$text}");
        }
        $integer = $m[1];
        $fraction = $m[2] ?? '';
        $exp = isset($m[3]) ? (int) $m[3] : 0;
        $digits = ltrim($integer . $fraction, '0');
        $leadingZeros = strlen($integer . $fraction) - strlen($digits);
        $digits = rtrim($digits, '0');
        if ($digits === '') {
            $digits = '0';
        }
        $exponent = strlen($integer) + $exp - $leadingZeros;

        return [$negative, $digits, $exponent];
    }

    /**
     * Positional decimal expansion without exponent.
     *
     * @return array{0: bool, 1: string, 2: string} negative, integer digits, fraction digits
     */
    public static function positional(float $value): array
    {
        if ($value == 0.0) {
            return [false, '0', ''];
        }
        [$negative, $digits, $exponent] = self::shortestDigits($value);
        $length = strlen($digits);
        if ($exponent <= 0) {
            $integer = '0';
            $fraction = str_repeat('0', -$exponent) . $digits;
        } elseif ($exponent >= $length) {
            $integer = $digits . str_repeat('0', $exponent - $length);
            $fraction = '';
        } else {
            $integer = substr($digits, 0, $exponent);
            $fraction = substr($digits, $exponent);
        }

        return [$negative, $integer, $fraction];
    }

    /**
     * FUN-22: rounds the positional digits half away from zero.
     *
     * @return array{0: bool, 1: string, 2: string}
     */
    public static function roundDecimal(float $value, int $decimals): array
    {
        [$negative, $integer, $fraction] = self::positional($value);
        if (strlen($fraction) <= $decimals) {
            return [$negative, $integer, $fraction . str_repeat('0', $decimals - strlen($fraction))];
        }
        $roundUp = ord($fraction[$decimals]) >= 0x35;
        $kept = $integer . substr($fraction, 0, $decimals);
        if ($roundUp) {
            $kept = self::increment($kept);
        }
        $splitAt = strlen($kept) - $decimals;
        $integerPart = substr($kept, 0, $splitAt);

        return [$negative, $integerPart === '' ? '0' : $integerPart, substr($kept, $splitAt)];
    }

    private static function increment(string $digits): string
    {
        $index = strlen($digits) - 1;
        while ($index >= 0) {
            if ($digits[$index] === '9') {
                $digits[$index] = '0';
                $index--;
            } else {
                $digits[$index] = chr(ord($digits[$index]) + 1);

                return $digits;
            }
        }

        return '1' . $digits;
    }

    /**
     * FUN-20 to FUN-24.
     */
    public static function format(float $value, int $decimals, string $dec, string $thousands): string
    {
        [$negative, $integer, $fraction] = self::roundDecimal($value, $decimals);
        $groups = [];
        while (strlen($integer) > 3) {
            array_unshift($groups, substr($integer, -3));
            $integer = substr($integer, 0, -3);
        }
        array_unshift($groups, $integer);
        $text = implode($thousands, $groups);
        if ($decimals > 0) {
            $text .= $dec . $fraction;
        }
        $allZero = preg_match('/^0*$/', implode('', $groups) . $fraction) === 1;

        return $negative && !$allZero ? '-' . $text : $text;
    }

    /**
     * FUN-25: the rounded value as a number.
     */
    public static function round(float $value, int $decimals): float
    {
        [$negative, $integer, $fraction] = self::roundDecimal($value, $decimals);
        $result = (float) ($integer . ($fraction !== '' ? '.' . $fraction : ''));

        return $negative && $result != 0.0 ? -$result : $result;
    }

    public static function isInteger(float $value): bool
    {
        return is_finite($value) && floor($value) == $value;
    }
}

File: packages/template-rust/src/value/number.rs
//! Number conversion and formatting (VAL-9, FUN-20 to FUN-25).

/// Largest safe integer magnitude (2^53 − 1).
pub const MAX_SAFE: f64 = 9007199254740991.0;

/// Shortest round-trip digits of a finite non-zero number: digits without leading or trailing
/// zeros, and the decimal exponent `n` such that value = 0.digits × 10^n.
pub fn shortest_digits(value: f64) -> (bool, String, i32) {
    let negative = value.is_sign_negative();
    let text = format!("{:e}", value.abs());
    let (mantissa, exponent) = text.split_once('e').expect("exponential form");
    let exponent: i32 = exponent.parse().expect("exponent");
    let mut digits: String = mantissa.chars().filter(|c| *c != '.').collect();
    while digits.len() > 1 && digits.ends_with('0') {
        digits.pop();
    }
    (negative, digits, exponent + 1)
}

/// VAL-9: ECMAScript Number::toString.
pub fn number_to_string(value: f64) -> String {
    if value == 0.0 {
        return "0".to_string();
    }
    let (negative, digits, n) = shortest_digits(value);
    let k = digits.len() as i32;
    let body = if k <= n && n <= 21 {
        format!("{digits}{}", "0".repeat((n - k) as usize))
    } else if 0 < n && n <= 21 {
        format!("{}.{}", &digits[..n as usize], &digits[n as usize..])
    } else if -6 < n && n <= 0 {
        format!("0.{}{digits}", "0".repeat((-n) as usize))
    } else {
        let e = n - 1;
        let sign = if e < 0 { '-' } else { '+' };
        if k == 1 {
            format!("{digits}e{sign}{}", e.abs())
        } else {
            format!("{}.{}e{sign}{}", &digits[..1], &digits[1..], e.abs())
        }
    };
    if negative { format!("-{body}") } else { body }
}

/// Positional decimal expansion without exponent.
pub fn positional(value: f64) -> (bool, String, String) {
    if value == 0.0 {
        return (false, "0".to_string(), String::new());
    }
    let (negative, digits, exponent) = shortest_digits(value);
    let length = digits.len() as i32;
    if exponent <= 0 {
        (negative, "0".to_string(), format!("{}{digits}", "0".repeat((-exponent) as usize)))
    } else if exponent >= length {
        (
            negative,
            format!("{digits}{}", "0".repeat((exponent - length) as usize)),
            String::new(),
        )
    } else {
        (
            negative,
            digits[..exponent as usize].to_string(),
            digits[exponent as usize..].to_string(),
        )
    }
}

fn increment_digits(digits: &str) -> String {
    let mut chars: Vec<u8> = digits.bytes().collect();
    let mut index = chars.len();
    while index > 0 {
        index -= 1;
        if chars[index] == b'9' {
            chars[index] = b'0';
        } else {
            chars[index] += 1;
            return String::from_utf8(chars).expect("digits");
        }
    }
    format!("1{}", String::from_utf8(chars).expect("digits"))
}

/// FUN-22: rounds the positional digits to `decimals` fraction digits, half away from zero.
pub fn round_decimal(value: f64, decimals: usize) -> (bool, String, String) {
    let (negative, integer, fraction) = positional(value);
    if fraction.len() <= decimals {
        return (negative, integer, format!("{fraction}{}", "0".repeat(decimals - fraction.len())));
    }
    let round_up = fraction.as_bytes()[decimals] >= b'5';
    let mut kept = format!("{integer}{}", &fraction[..decimals]);
    if round_up {
        kept = increment_digits(&kept);
    }
    let split = kept.len() - decimals;
    let integer_part = if split == 0 { "0".to_string() } else { kept[..split].to_string() };
    (negative, integer_part, kept[split..].to_string())
}

/// FUN-20 to FUN-24.
pub fn format_number(value: f64, decimals: usize, dec: &str, thousands: &str) -> String {
    let (negative, integer, fraction) = round_decimal(value, decimals);
    let mut groups: Vec<String> = Vec::new();
    let mut rest = integer.as_str();
    while rest.len() > 3 {
        groups.insert(0, rest[rest.len() - 3..].to_string());
        rest = &rest[..rest.len() - 3];
    }
    groups.insert(0, rest.to_string());
    let mut text = groups.join(thousands);
    if decimals > 0 {
        text.push_str(dec);
        text.push_str(&fraction);
    }
    let all_zero = integer.bytes().chain(fraction.bytes()).all(|b| b == b'0');
    if negative && !all_zero { format!("-{text}") } else { text }
}

/// FUN-25: the rounded value as a number.
pub fn round_number(value: f64, decimals: usize) -> f64 {
    let (negative, integer, fraction) = round_decimal(value, decimals);
    let text = if fraction.is_empty() {
        integer
    } else {
        format!("{integer}.{fraction}")
    };
    let result: f64 = text.parse().unwrap_or(0.0);
    if negative && result != 0.0 { -result } else { result }
}
